enhancements/sig-architecture/32-extensiblescheduling/32-extensiblescheduling.md

```diff
@@ -0,0 +1,363 @@
+
+The code change includes add PreScore() in Prioritizer interface, in which to create ManagedClusterScalar CR before Score().
+```go
+// Prioritizer defines a prioritizer plugin that score each cluster. The score is normalized
+// as a floating betwween 0 and 1.
+type Prioritizer interface {
+	Plugin
+
+	// PreScore() do some prepare work before Score().
+	// For example, creating ManagedClusterScalar CR before Score().
+	PreScore(ctx context.Context, placement *clusterapiv1alpha1.Placement, clusters []*clusterapiv1.ManagedCluster) error
+
+	// Score gives the score to a list of the clusters, it returns a map with the key as
+	// the cluster name.
+	Score(ctx context.Context, placement *clusterapiv1alpha1.Placement, clusters []*clusterapiv1.ManagedCluster) (map[string]int64, error)
+}
+```
+
+- Delete
+
+The deletion of ManagedClusterScalar CR is handled by the framework (out of all the prioritizer plugins). The ManagedClusterScalar CR is created by each prioritizer plugin on demand, and the CR should be shared by some placements cross namespaces. When the ManagedClusterScalar CR is not used by any placement, need to clean it to avoid unnecessary api call.
+
+There will be a GC() funcion in the framework, being called every 10 minutes. Go through all the Placment CRs and ManagedClusterScalar CRs, and clean not used ManagedClusterScalar CRs.
+
+- Update
+
+ManagedClusterScalar CR's status is updated by customized agent or third party controllers. Out of scope and won't be covered in this proposal.
+
+### Let placement prioritizer support rating clusters with the scalar value of the CR.
+Once ManagedClusterScalar CR is created in PreScore(), a customized agent or third party controllers should watch the change and update the CR status as below.
+```yaml
+apiVersion: cluster.open-cluster-management.io/v1alpha1
+kind: ManagedClusterScalar
+metadata:
+  name: {PrioritizerName}
+  namespace: cluster1
+spec:
+  prioritizerName: {PrioritizerNameInPlacement}
+status:
+  conditions:
+  - lastTransitionTime: "2021-10-28T08:31:39Z"
+    message: ManagedClusterScalar updated successfully
+    reason: ManagedClusterScalarUpdated
+    status: "True"
+    type: ManagedClusterScalarUpdated
+  validUntil: "2021-10-29T18:31:39Z"
+  scalar: 74
+```
+Usually each ManagedClusterScalar CR will be created in more than one cluster namespace, different clusters update the status separately. So the prioritizer plugin which depends on ManagedClusterScalar should follow below logic to get avaliable scalar and sort all the clusters.
+- Prescore
+
+In prescore stage, prioritizer plugin should create the CR before use it (have talked about this in previous section).
+The plugin should also check if 80% of the CRs have a valid scalar (80% is a reasonable threshold hardcode so far, might make it configurable in the future). If yes, the plugin will continue to score the clusters. If not, terminate this scheduling and wait for the next reschedule.
```

Comment:
is it really better to place nothing instead of placing what you can?

It seems like you have two cases
1. a placement contains a managedcluster that no longer has an addonPlacementScore.
2. a placement needs to select additional managedclusters and the only ones available have no addonPlacementScore

It seems like always collecting the data if the managedcluster has the addon installed will reduce the mental load around, "what do we do if we don't have this information".

Reply:
Will update the doc to handle the 2 cases.